sw/device/lib/crypto/aes_gcm/aes_gcm.c

```diff
@@ -0,0 +1,356 @@
+ * @param out Buffer for output
+ */
+static inline void galois_mulx(const aes_block_t *p, aes_block_t *out) {
+  // Get the very rightmost bit of the input block (coefficient of x^127).
+  uint8_t p127 = block_byte_get(p, kAesBlockNumBytes - 1) & 1;
+  // Set the output to p >> 1.
+  memcpy(out->data, p->data, kAesBlockNumBytes);
+  block_shiftr(out, 1);
+  // If the highest coefficient was 1, then subtract the polynomial that
+  // corresponds to (modulus - 2^128).
+  uint32_t mask = 0 - p127;
+  out->data[0] ^= (0xe1 & mask);
+}
+
+/**
+ * Construct a product table for the hash subkey H.
+ *
+ * The product table entry at index i is equal to i * H, where both i and H are
+ * interpreted as Galois field polynomials.
+ *
+ * @param H Hash subkey
+ * @param product_table Buffer for output
+ */
+static void make_product_table(const aes_block_t *H,
+                               aes_gcm_product_table_t *tbl) {
+  // Initialize 0 * H = 0.
+  memset(tbl->products[0].data, 0, kAesBlockNumBytes);
+  // Initialize 1 * H = H.
+  memcpy(tbl->products[0x8].data, H->data, kAesBlockNumBytes);
+
+  // To get remaining entries, we use a variant of "shift and add"; in
+  // polynomial terms, a shift is a multiplication by x. Note that, because the
+  // processor represents bytes with the MSB on the left and NIST uses a fully
+  // little-endian polynomial representation with the MSB on the right, we have
+  // to reverse the bits of the indices.
+  for (size_t i = 2; i < 16; i += 2) {
+    // Find the product corresponding to (i >> 1) * H and multiply by x to
+    // shift 1; this will be i * H.
+    galois_mulx(&tbl->products[reverse_bits(i >> 1)],
+                &tbl->products[reverse_bits(i)]);
+
+    // Add H to i * H to get (i + 1) * H.
+    block_xor(&tbl->products[reverse_bits(i)], H,
+              &tbl->products[reverse_bits(i + 1)]);
+  }
+}
+
+/**
+ * Computes the "product block" as defined in NIST SP800-38D, section 6.3.
+ *
+ * The NIST documentation shows a very slow double-and-add algorithm, which
+ * iterates through the first operand bit-by-bit. However, since the second
+ * operand is always the hash subkey H, we can speed things up significantly
+ * with a little precomputation.
+ *
+ * The `H_table` input should be a precomputed table with the products of H and
+ * 256 all possible byte values (see `make_product_table`).
+ *
+ * This operation corresponds to multiplication in the Galois field with order
+ * 2^128, modulo the polynomial x^128 +  x^8 + x^2 + x + 1
+ *
+ * @param p Polynomial to multiply
+ * @param H_table Precomputed product table for the hash subkey
+ * @param result Block in which to store output
+ */
+static void galois_mul(const aes_block_t *p, const aes_gcm_product_table_t *tbl,
+                       aes_block_t *result) {
+  // Initialize the result to 0.
+  memset(result->data, 0, kAesBlockNumBytes);
+
+  // 4-bit windows, so number of windows is 2x number of bytes.
+  size_t nwindows = kAesBlockNumBytes << 1;
```

Comment:
Nit: Could this be `const size_t kNumWindows = ...`? I wish C had constexpr.

Reply:
(My intent with this suggestion is to clarify on line 259 that that `nwindows` is not a parameter of `galois_mul()`.)

Reply:
Done!